File: modules/xml/app/model/CorrectionErrorHandler.java
package model;

import java.util.LinkedList;
import java.util.List;

import org.xml.sax.ErrorHandler;
import org.xml.sax.SAXException;
import org.xml.sax.SAXParseException;

public class CorrectionErrorHandler implements ErrorHandler {

  private List<XmlError> output = new LinkedList<>();

  @Override
  public void error(SAXParseException exception) throws SAXException {
    output.add(new XmlError(exception, XmlErrorType.ERROR));
  }

  @Override
  public void fatalError(SAXParseException exception) throws SAXException {
    output.add(new XmlError(exception, XmlErrorType.FATALERROR));
  }

  public List<XmlError> getErrors() {
    return output;
  }

  @Override
  public void warning(SAXParseException exception) throws SAXException {
    output.add(new XmlError(exception, XmlErrorType.WARNING));
  }
}

File: modules/xml/app/model/CorrectionErrorHandler.scala
package model

import java.io.IOException

import scala.collection.mutable.ListBuffer

import org.xml.sax.{ ErrorHandler, SAXParseException }

import model.result.{ EvaluationResult, SuccessType }

abstract sealed class XmlError(val title: String, val errorMessage: String, val line: Int, s: SuccessType)
  extends EvaluationResult(s) {
  val lineStr = if (line != -1) s" in Zeile $line" else ""
}

case class FatalXmlError(e: SAXParseException) extends XmlError("Fataler Fehler", e.getMessage, e.getLineNumber, SuccessType.NONE)

case class ErrorXmlError(e: SAXParseException) extends XmlError("Fehler", e.getMessage, e.getLineNumber, SuccessType.NONE)

case class WarningXmlError(e: SAXParseException) extends XmlError("Warnung", e.getMessage, e.getLineNumber, SuccessType.PARTIALLY)

case class FailureXmlError(msg: String, error: IOException = null) extends XmlError("Fehlschlag", msg, -1, SuccessType.NONE)

class CorrectionErrorHandler extends ErrorHandler {

  val errors: ListBuffer[XmlError] = ListBuffer.empty

  override def error(exception: SAXParseException) = errors += ErrorXmlError(exception)

  override def fatalError(exception: SAXParseException) = errors += FatalXmlError(exception)

  override def warning(exception: SAXParseException) = errors += WarningXmlError(exception)

}
